File: tests/cross_driver/java/src/main/java/com/secantus/smokes/DropAllUsersSmoke.java
package com.secantus.smokes;

import com.mongodb.ConnectionString;
import com.mongodb.MongoClientSettings;
import com.mongodb.MongoCredential;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import org.bson.Document;

import java.util.List;

/**
 * Cross-driver dropAllUsersFromDatabase smoke — Java.
 * <p>
 * Provisions two users in {@code shop} and one in {@code other}, then
 * runs {@code dropAllUsersFromDatabase} on {@code shop}. Asserts
 * {@code n: 2} (only shop users removed), {@code shop} now has zero
 * users, and {@code other} still has Carol.
 */
public final class DropAllUsersSmoke {

    private DropAllUsersSmoke() {}

    public static void main(String[] args) {
        String uri = System.getenv("MONGODB_URI");
        String adminPwd = System.getenv("ADMIN_PASSWORD");
        if (uri == null || adminPwd == null) {
            System.err.println("MONGODB_URI and ADMIN_PASSWORD required");
            System.exit(2);
        }

        try (MongoClient root = clientFor(uri, "root", adminPwd, "admin")) {
            root.getDatabase("shop").runCommand(new Document()
                .append("createUser", "alice").append("pwd", "p")
                .append("roles", List.of(new Document()
                    .append("role", "read").append("db", "shop"))));
            root.getDatabase("shop").runCommand(new Document()
                .append("createUser", "bob").append("pwd", "p")
                .append("roles", List.of(new Document()
                    .append("role", "readWrite").append("db", "shop"))));
            root.getDatabase("other").runCommand(new Document()
                .append("createUser", "carol").append("pwd", "p")
                .append("roles", List.of(new Document()
                    .append("role", "read").append("db", "other"))));

            Document res = root.getDatabase("shop").runCommand(new Document(
                "dropAllUsersFromDatabase", 1));
            int n = res.getInteger("n", -1);
            if (n != 2) {
                System.err.printf("n: got %d, want 2%n", n);
                System.exit(1);
            }

            Document shopUsersRes = root.getDatabase("shop").runCommand(
                new Document("usersInfo", 1));
            List<?> shopUsers = (List<?>) shopUsersRes.get("users");
            if (shopUsers == null || !shopUsers.isEmpty()) {
                System.err.printf("shop users: %s%n", shopUsers);
                System.exit(1);
            }

            Document otherUsersRes = root.getDatabase("other").runCommand(
                new Document("usersInfo", 1));
            @SuppressWarnings("unchecked")
            List<Document> otherUsers = (List<Document>) otherUsersRes.get("users");
            if (otherUsers == null || otherUsers.size() != 1
                    || !"carol".equals(otherUsers.get(0).getString("user"))) {
                System.err.printf("other users: %s%n", otherUsers);
                System.exit(1);
            }

            System.out.println("OK");
        }
    }

    private static MongoClient clientFor(String uri, String user, String pwd, String authSource) {
        MongoCredential credential = MongoCredential.createScramSha256Credential(
            user, authSource, pwd.toCharArray());
        MongoClientSettings settings = MongoClientSettings.builder()
            .applyConnectionString(new ConnectionString(uri))
            .credential(credential)
            .applyToClusterSettings(b -> b.serverSelectionTimeout(5,
                java.util.concurrent.TimeUnit.SECONDS))
            .build();
        return MongoClients.create(settings);
    }
}

File: tests/cross_driver/node/drop_all_users_smoke.js
// Cross-driver dropAllUsersFromDatabase smoke — Node.

const { MongoClient } = require('mongodb');

function fail(msg) {
  console.error(msg);
  process.exit(1);
}

function clientFor(uri, user, pwd, authSource) {
  return new MongoClient(uri, {
    auth: { username: user, password: pwd },
    authSource,
    authMechanism: 'SCRAM-SHA-256',
    serverSelectionTimeoutMS: 5000,
  });
}

async function main() {
  const uri = process.env.MONGODB_URI;
  const adminPwd = process.env.ADMIN_PASSWORD;
  if (!uri || !adminPwd) fail('MONGODB_URI and ADMIN_PASSWORD required');

  const root = clientFor(uri, 'root', adminPwd, 'admin');
  await root.connect();
  try {
    await root
      .db('shop')
      .command({
        createUser: 'alice',
        pwd: 'p',
        roles: [{ role: 'read', db: 'shop' }],
      });
    await root
      .db('shop')
      .command({
        createUser: 'bob',
        pwd: 'p',
        roles: [{ role: 'readWrite', db: 'shop' }],
      });
    await root
      .db('other')
      .command({
        createUser: 'carol',
        pwd: 'p',
        roles: [{ role: 'read', db: 'other' }],
      });

    const res = await root.db('shop').command({ dropAllUsersFromDatabase: 1 });
    if (res.n !== 2) fail(`n: got ${res.n}, want 2`);

    const shopUsers = (await root.db('shop').command({ usersInfo: 1 })).users;
    if (shopUsers.length !== 0) fail(`shop users: ${JSON.stringify(shopUsers)}`);

    const otherUsers = (await root.db('other').command({ usersInfo: 1 })).users;
    if (otherUsers.length !== 1 || otherUsers[0].user !== 'carol') {
      fail(`other users: ${JSON.stringify(otherUsers)}`);
    }

    console.log('OK');
  } finally {
    await root.close();
  }
}

main().catch((err) => fail(`uncaught: ${err && err.stack ? err.stack : err}`));
